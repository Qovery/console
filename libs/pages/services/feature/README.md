# pages-services-feature

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test pages-services-feature` to execute the unit tests via [Jest](https://jestjs.io).
