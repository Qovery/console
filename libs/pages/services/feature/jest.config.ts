/* eslint-disable */
export default {
  displayName: 'pages-services-feature',
  preset: '../../../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../../coverage/libs/pages/services/feature',
}
