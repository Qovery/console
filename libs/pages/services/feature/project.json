{
  "root": "libs/pages/services/feature",
  "sourceRoot": "libs/pages/services/feature/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/pages/services/feature/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/pages/services/feature"],
      "options": {
        "jestConfig": "libs/pages/services/feature/jest.config.js",
        "passWithNoTests": true
      }
    }
  }
}
