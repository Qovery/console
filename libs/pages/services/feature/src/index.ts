export * from './lib/services-page'
export * from './lib/components/deployments/deployments'
export * from './lib/components/general/general'
export * from './lib/router/router'
