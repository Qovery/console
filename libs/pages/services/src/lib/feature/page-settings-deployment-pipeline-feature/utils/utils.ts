import { type DeploymentStageResponse, type DeploymentStageServiceResponse } from 'qovery-typescript-axios'
import { type DraggableLocation } from 'react-beautiful-dnd'

export const reorderService = (
  stages: DeploymentStageResponse[],
  destinationIndex: number,
  startIndex: number,
  endIndex: number
): DeploymentStageResponse[] => {
  const cloneStages = [...stages]
  const currentStage = cloneStages[destinationIndex]
  const currentStageServices = [...(currentStage.services as DeploymentStageServiceResponse[])]

  const [removed] = currentStageServices.splice(startIndex, 1)
  currentStageServices.splice(endIndex, 0, removed)

  // add reorder on the clone stages array
  const result = cloneStages.map((stage, index) => {
    if (index === destinationIndex) {
      return { ...stage, services: currentStageServices }
    }
    return stage
  })

  return result
}

export const move = (
  stages: DeploymentStageResponse[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation
): DeploymentStageResponse[] | undefined => {
  let result
  const cloneStages = [...stages]

  const sourceDroppableId = parseInt(droppableSource.droppableId, 10)
  const destinationDroppableId = parseInt(droppableDestination.droppableId, 10)

  // find the source stage
  const sourceStage = cloneStages.find((stage, index) => index === sourceDroppableId)

  if (sourceStage) {
    // find the destination stage
    const destinationStage = cloneStages.find((stage, index) => index === destinationDroppableId)

    if (destinationStage?.services && sourceStage?.services) {
      const cloneSourceServices = [...sourceStage.services]
      // remove the service from the source services
      const [serviceMoved] = cloneSourceServices.splice(droppableSource.index, 1)

      // add the service on the destination services on the droppable destination index
      const cloneDestinationServices = [...destinationStage.services]
      cloneDestinationServices.splice(droppableDestination.index, 0, serviceMoved)

      // add new services for source and destination on the stages
      result = cloneStages.map((stage, index) => {
        if (index === sourceDroppableId) {
          return { ...stage, services: cloneSourceServices }
        }
        if (index === destinationDroppableId) {
          return { ...stage, services: cloneDestinationServices }
        }
        return stage
      })
    }
  }

  return result
}
