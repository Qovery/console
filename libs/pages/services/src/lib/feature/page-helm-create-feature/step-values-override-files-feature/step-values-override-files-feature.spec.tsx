import { wrapWithReactHookForm } from '__tests__/utils/wrap-with-react-hook-form'
import { useForm } from 'react-hook-form'
import { type HelmValuesFileData } from '@qovery/domains/service-helm/feature'
import { renderHook, renderWithProviders, screen, waitFor } from '@qovery/shared/util-tests'
import { type HelmGeneralData } from '../page-helm-create-feature'
import { HelmCreateContext } from '../page-helm-create-feature'
import StepValuesOverrideFilesFeature from './step-values-override-files-feature'

describe('StepValuesOverrideFilesFeature', () => {
  it('should render successfully', () => {
    const { result: generalForm } = renderHook(() =>
      useForm<HelmGeneralData>({
        mode: 'onChange',
        defaultValues: {
          source_provider: 'HELM_REPOSITORY',
          repository: 'https://charts.bitnami.com/bitnami',
          chart_name: 'nginx',
          chart_version: '8.9.0',
          arguments: '',
        },
      })
    )

    const { result: valuesOverrideFileForm } = renderHook(() =>
      useForm<HelmValuesFileData>({
        mode: 'onChange',
        defaultValues: {
          type: 'NONE',
        },
      })
    )

    const { baseElement } = renderWithProviders(
      <HelmCreateContext.Provider
        value={{
          currentStep: 1,
          setCurrentStep: jest.fn(),
          generalForm: generalForm.current,
          valuesOverrideFileForm: valuesOverrideFileForm.current,
        }}
      >
        <StepValuesOverrideFilesFeature />
      </HelmCreateContext.Provider>
    )
    expect(baseElement).toBeTruthy()
  })

  it('should submit a form with a git repository', async () => {
    const { result: generalForm } = renderHook(() =>
      useForm<HelmGeneralData>({
        mode: 'onChange',
        defaultValues: {
          source_provider: 'GIT',
          provider: 'GITHUB',
          repository: 'Qovery/github',
          branch: 'main',
          root_path: '/',
        },
      })
    )

    const { result: valuesOverrideFileForm } = renderHook(() =>
      useForm<HelmValuesFileData>({
        mode: 'onChange',
        defaultValues: {
          type: 'GIT_REPOSITORY',
          provider: 'GITHUB',
          repository: 'Qovery/github',
          branch: 'main',
          paths: '/',
        },
      })
    )

    const { baseElement, userEvent } = renderWithProviders(
      wrapWithReactHookForm(
        <HelmCreateContext.Provider
          value={{
            currentStep: 2,
            setCurrentStep: jest.fn(),
            generalForm: generalForm.current,
            valuesOverrideFileForm: valuesOverrideFileForm.current,
          }}
        >
          <StepValuesOverrideFilesFeature />
        </HelmCreateContext.Provider>
      )
    )

    const button = screen.getByRole('button', { name: 'Continue' })

    // wait for form to be valid because we have selects (necessary with react hook form)
    waitFor(async () => {
      expect(button).toBeEnabled()
      await userEvent.click(button)
    })

    expect(baseElement).toMatchSnapshot()
  })

  it('should submit a form with a yaml', async () => {
    const { result: generalForm } = renderHook(() =>
      useForm<HelmGeneralData>({
        mode: 'onChange',
        defaultValues: {
          source_provider: 'GIT',
          provider: 'GITHUB',
          repository: 'Qovery/github',
          branch: 'main',
          root_path: '/',
        },
      })
    )

    const { result: valuesOverrideFileForm } = renderHook(() =>
      useForm<HelmValuesFileData>({
        mode: 'onChange',
        defaultValues: {
          type: 'YAML',
          content: 'test',
        },
      })
    )

    const { baseElement, userEvent } = renderWithProviders(
      wrapWithReactHookForm(
        <HelmCreateContext.Provider
          value={{
            currentStep: 2,
            setCurrentStep: jest.fn(),
            generalForm: generalForm.current,
            valuesOverrideFileForm: valuesOverrideFileForm.current,
          }}
        >
          <StepValuesOverrideFilesFeature />
        </HelmCreateContext.Provider>
      )
    )

    const button = screen.getByRole('button', { name: 'Continue' })

    expect(button).toBeEnabled()
    await userEvent.click(button)

    expect(baseElement).toMatchSnapshot()
  })
})
