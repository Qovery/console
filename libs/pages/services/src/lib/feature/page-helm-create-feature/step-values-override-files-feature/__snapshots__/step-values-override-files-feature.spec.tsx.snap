// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepValuesOverrideFilesFeature should submit a form with a git repository 1`] = `
<body>
  <div />
  <div />
  <div>
    <div
      class="absolute h-full bg-white w-full pointer-events-none"
      style="left: -30%;"
    />
    <div
      class="flex w-full overflow-auto"
    >
      <section
        class="bg-white pt-14 w-full lg:w-[70%]"
      >
        <div
          class="px-12 mx-auto relative pb-14 max-w-content-with-navigation-left"
          data-testid="funnel-body-content"
        >
          <section
            class="flex flex-col items-start"
          >
            <h1
              class="text-neutral-400 text-xl font-bold mb-2"
            >
              Values override as file
            </h1>
            <p
              class="text-sm text-neutral-350 mb-2"
            >
              Define the YAML file(s) to be applied as override to the default values.yaml delivered with the chart. It is highly recommended to store the override file(s) in a git repository. To get all the Qovery functionalities, add the macro “qovery.labels.service” and "qovery.annotations.service" within the field managing the labels/annotations assigned to the deployed Pods/Deployments/Services/Jobs.
            </p>
            <span
              aria-controls="radix-:rd:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="text-sm cursor-pointer text-brand-500 hover:text-brand-600 transition font-medium mb-6"
              data-state="closed"
              type="button"
            >
              How it works 
              <span
                class="icon-regular-circle-question shrink-0 text-xs"
                role="img"
              />
            </span>
            <div
              class="flex flex-row items-baseline gap-x-3 p-3 border rounded border-yellow-600 bg-yellow-50 text-yellow-700 mb-6 text-xs"
              data-accent-color="yellow"
            >
              <div
                class="-order-2 text-base"
              >
                <span
                  class="icon-solid-triangle-exclamation shrink-0 "
                  role="img"
                />
              </div>
              <div
                class="-order-1 mr-auto gap-x-3 text-neutral-400"
              >
                <p
                  class="font-medium mb-1"
                >
                  Add the Qovery macros to your override
                </p>
                <p
                  class="mb-1"
                >
                  To get all the Qovery functionalities (Logs, Statuses, Helm stop and restart), add the macro “qovery.labels.service” and "qovery.annotations.service" within the field managing the labels/annotations assigned to the deployed Pods/Deployments/Services/Jobs.
                </p>
                <a
                  class="cursor-pointer transition duration-100 font-medium inline-flex items-center gap-1 text-sky-500 hover:text-sky-600 text-xs"
                  href="https://hub.qovery.com/docs/using-qovery/configuration/helm/#values"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Click here for more details
                  <span
                    class="icon-solid-arrow-up-right-from-square shrink-0 text-2xs leading-4"
                    role="img"
                  />
                </a>
              </div>
            </div>
            <form
              class="w-full"
            >
              <div
                class=""
              >
                <div
                  class="input input--select    "
                  data-testid="select"
                >
                  <label
                    class="!text-xs !translate-y-0"
                    for="File source"
                  >
                    File source
                  </label>
                  <div
                    class=" css-b62m3t-container"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-3-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="input-select__control css-13cymwt-control"
                    >
                      <div
                        class="input-select__value-container input-select__value-container--has-value css-1fdsijx-ValueContainer"
                      >
                        <span
                          class="text-sm text-neutral-400 mr-1"
                        >
                          Git repository
                        </span>
                        <input
                          aria-autocomplete="list"
                          aria-expanded="false"
                          aria-haspopup="true"
                          aria-readonly="true"
                          class="css-mohuvp-dummyInput-DummyInput"
                          id="File source"
                          inputmode="none"
                          role="combobox"
                          tabindex="0"
                          value=""
                        />
                      </div>
                      <div
                        class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                    <input
                      name="File source"
                      type="hidden"
                      value="GIT_REPOSITORY"
                    />
                  </div>
                  <input
                    name="File source"
                    type="hidden"
                    value="GIT_REPOSITORY"
                  />
                  <div
                    class="absolute top-1/2 -translate-y-1/2 right-4 pointer-events-none"
                  >
                    <span
                      class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
                      role="img"
                    />
                  </div>
                </div>
              </div>
              <section
                class="flex flex-col"
              >
                <h2
                  class="text-neutral-400 text-base font-bold mt-10 mb-2"
                >
                  Override from repository
                </h2>
                <p
                  class="text-sm text-neutral-350 mb-6"
                >
                  Specify the repository and the path containing the override yaml file to be passed via the “-f” helm argument. More than one file can be used as override by adding them in the path field separated by a semi-colon. If you don’t have a repository, you can set the override manually or via a raw YAML file.
                </p>
                <div
                  class="flex flex-col gap-3"
                >
                  <div
                    class=""
                  >
                    <div
                      class="input input--select    "
                      data-testid="select"
                    >
                      <label
                        class="text-sm translate-y-2 top-1.5"
                        for="Git repository"
                      >
                        Git repository
                      </label>
                      <div
                        class=" css-b62m3t-container"
                      >
                        <span
                          class="css-1f43avz-a11yText-A11yText"
                          id="react-select-4-live-region"
                        />
                        <span
                          aria-atomic="false"
                          aria-live="polite"
                          aria-relevant="additions text"
                          class="css-1f43avz-a11yText-A11yText"
                        />
                        <div
                          class="input-select__control css-13cymwt-control"
                        >
                          <div
                            class="input-select__value-container css-1fdsijx-ValueContainer"
                          >
                            <div
                              class="input-select__placeholder css-1jqq78o-placeholder"
                              id="react-select-4-placeholder"
                            >
                              Select...
                            </div>
                            <input
                              aria-autocomplete="list"
                              aria-describedby="react-select-4-placeholder"
                              aria-expanded="false"
                              aria-haspopup="true"
                              aria-readonly="true"
                              class="css-mohuvp-dummyInput-DummyInput"
                              id="Git repository"
                              inputmode="none"
                              role="combobox"
                              tabindex="0"
                              value=""
                            />
                          </div>
                          <div
                            class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
                          >
                            <span
                              class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                            />
                            <div
                              aria-hidden="true"
                              class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                            >
                              <svg
                                aria-hidden="true"
                                class="css-tj5bde-Svg"
                                focusable="false"
                                height="20"
                                viewBox="0 0 20 20"
                                width="20"
                              >
                                <path
                                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                                />
                              </svg>
                            </div>
                          </div>
                        </div>
                        <input
                          name="Git repository"
                          type="hidden"
                          value=""
                        />
                      </div>
                      <input
                        name="Git repository"
                        type="hidden"
                        value=""
                      />
                      <div
                        class="absolute top-1/2 -translate-y-1/2 right-4 pointer-events-none"
                      >
                        <span
                          class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
                          role="img"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="flex justify-center"
                  >
                    <div
                      class="aspect-square border-2  rounded-full border-solid border-r-neutral-500 animate-spin w-4 "
                      data-testid="spinner"
                    />
                  </div>
                  <div
                    class="flex justify-center"
                  >
                    <div
                      class="aspect-square border-2  rounded-full border-solid border-r-neutral-500 animate-spin w-4 "
                      data-testid="spinner"
                    />
                  </div>
                  <div>
                    <div
                      class=""
                      data-testid="input-text-wrapper"
                    >
                      <div
                        class="relative"
                      >
                        <div
                          aria-label="input-container"
                          class="input    input--label-up "
                        >
                          <div
                            class=""
                          >
                            <label
                              class="text-sm translate-y-2"
                              for="Overrides path"
                            >
                              Overrides path
                            </label>
                            <input
                              class="input__value "
                              data-testid="input-text"
                              id="Overrides path"
                              name="paths"
                              type="text"
                              value="/"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                    <p
                      class="text-xs text-neutral-350 ml-4 mt-1"
                    >
                      Specify multiple paths by separating them with a semi-colon
                    </p>
                  </div>
                </div>
              </section>
              <div
                class="flex justify-between mt-10"
              >
                <button
                  class="inline-flex items-center shrink-0 font-medium transition text-sm h-10 px-4 rounded border border-neutral-250 dark:border-neutral-350 text-neutral-400 dark:text-neutral-300 hover:[&:not(:active)]:border-neutral-300 active:bg-neutral-150 disabled:text-neutral-300 disabled:bg-neutral-150 disabled:border-none bg-neutral-100 dark:bg-neutral-500 dark:hover:bg-neutral-550"
                  type="button"
                >
                  Back
                </button>
                <div
                  class="flex gap-3"
                >
                  <button
                    class="inline-flex items-center shrink-0 font-medium transition text-sm h-10 px-4 rounded bg-brand-500 hover:bg-brand-600 text-white disabled:text-brand-300 disabled:bg-brand-100 disabled:border-none"
                    type="submit"
                  >
                    Continue
                  </button>
                </div>
              </div>
            </form>
          </section>
        </div>
      </section>
      <aside
        class="w-[30%] sticky top-0 pt-14 pl-10 hidden lg:block"
      >
        <div
          class="bg-neutral-50 p-8 border-neutral-200 border-b border-l rounded max-w-[22.5rem] "
          data-testid="funnel-body-help"
        >
          <section
            class="flex flex-col"
          >
            <svg
              class="shrink-0 w-7 h-7 mb-4"
              fill="none"
              height="28"
              name="INFORMATION"
              viewBox="0 0 28 28"
              width="1.5rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="#B160F0"
                height="28"
                rx="3"
                width="28"
              />
              <path
                d="M11.813 19.191c0 .165.027.329.136.465l.465.711c.137.219.465.383.738.383h1.668c.274 0 .602-.164.739-.383l.464-.71c.082-.11.137-.329.137-.466l.027-1.066h-4.375v1.066zm2.16-12.414c-2.79 0-4.786 2.243-4.786 4.785 0 1.204.438 2.297 1.176 3.145.465.52 1.149 1.613 1.422 2.516v.027h4.402v-.027c.274-.903.957-1.996 1.422-2.516a4.673 4.673 0 001.203-3.145c0-2.652-2.187-4.812-4.84-4.785zM14 9.402c-1.23 0-2.188.957-2.188 2.16a.45.45 0 01-.437.438.432.432 0 01-.438-.438c0-1.667 1.368-3.035 3.063-3.035a.47.47 0 01.438.438.45.45 0 01-.438.437z"
                fill="#fff"
              />
            </svg>
            <h2
              class="text-neutral-400 text-base font-bold mb-4"
            >
              Using values overrides
            </h2>
            <ul
              class="list-disc pl-4 text-neutral-400 text-sm mb-8"
            >
              <li
                class="mb-3"
              >
                Your helm chart might have already a variables.yaml file with some basic configuration. In this section you can define your own overrides to customize the helm chart behaviour.
              </li>
              <li
                class="mb-3"
              >
                You can define the overrides by selecting a YAML file from a git repository (preferred) or by passing a raw YAML file.
              </li>
              <li
                class="mb-3"
              >
                You can use the Qovery environment variables as overrides by using the placeholder "qovery.env.ENV_VAR_NAME” (Example: qovery.env.DB_URL. Qovery will manage the replacement of those placeholders at deployment time.
              </li>
              <li
                class="mb-3"
              >
                To get all the Qovery functionalities, add the macro “qovery.labels.service” and "qovery.annotations.service" within the field managing the labels/annotations assigned to the deployed Pods/Deployments/Services/Jobs. Qovery will automatically replace the macro with a dedicated set of labels/annotations.
              </li>
            </ul>
            <div
              class="py-8 border-t border-neutral-200 -mx-8 px-8 pb-0"
              data-testid="help-section"
            >
              <p
                class="text-neutral-400 text-sm mb-5"
              >
                Need help? You may find these links useful
              </p>
              <div
                class="flex flex-col gap-2"
              >
                <a
                  class="cursor-pointer transition duration-100 font-medium items-center gap-1 text-sky-500 hover:text-sky-600 text-sm flex"
                  href="https://hub.qovery.com/docs/using-qovery/configuration/helm/"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  How to configure my Helm chart
                  <span
                    class="icon-solid-arrow-up-right-from-square shrink-0 text-xs leading-5"
                    role="img"
                  />
                </a>
              </div>
            </div>
          </section>
        </div>
      </aside>
    </div>
  </div>
</body>
`;

exports[`StepValuesOverrideFilesFeature should submit a form with a yaml 1`] = `
<body>
  <div />
  <div />
  <div>
    <div
      class="absolute h-full bg-white w-full pointer-events-none"
      style="left: -30%;"
    />
    <div
      class="flex w-full overflow-auto"
    >
      <section
        class="bg-white pt-14 w-full lg:w-[70%]"
      >
        <div
          class="px-12 mx-auto relative pb-14 max-w-content-with-navigation-left"
          data-testid="funnel-body-content"
        >
          <section
            class="flex flex-col items-start"
          >
            <h1
              class="text-neutral-400 text-xl font-bold mb-2"
            >
              Values override as file
            </h1>
            <p
              class="text-sm text-neutral-350 mb-2"
            >
              Define the YAML file(s) to be applied as override to the default values.yaml delivered with the chart. It is highly recommended to store the override file(s) in a git repository. To get all the Qovery functionalities, add the macro “qovery.labels.service” and "qovery.annotations.service" within the field managing the labels/annotations assigned to the deployed Pods/Deployments/Services/Jobs.
            </p>
            <span
              aria-controls="radix-:rk:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="text-sm cursor-pointer text-brand-500 hover:text-brand-600 transition font-medium mb-6"
              data-state="closed"
              type="button"
            >
              How it works 
              <span
                class="icon-regular-circle-question shrink-0 text-xs"
                role="img"
              />
            </span>
            <div
              class="flex flex-row items-baseline gap-x-3 p-3 border rounded border-yellow-600 bg-yellow-50 text-yellow-700 mb-6 text-xs"
              data-accent-color="yellow"
            >
              <div
                class="-order-2 text-base"
              >
                <span
                  class="icon-solid-triangle-exclamation shrink-0 "
                  role="img"
                />
              </div>
              <div
                class="-order-1 mr-auto gap-x-3 text-neutral-400"
              >
                <p
                  class="font-medium mb-1"
                >
                  Add the Qovery macros to your override
                </p>
                <p
                  class="mb-1"
                >
                  To get all the Qovery functionalities (Logs, Statuses, Helm stop and restart), add the macro “qovery.labels.service” and "qovery.annotations.service" within the field managing the labels/annotations assigned to the deployed Pods/Deployments/Services/Jobs.
                </p>
                <a
                  class="cursor-pointer transition duration-100 font-medium inline-flex items-center gap-1 text-sky-500 hover:text-sky-600 text-xs"
                  href="https://hub.qovery.com/docs/using-qovery/configuration/helm/#values"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Click here for more details
                  <span
                    class="icon-solid-arrow-up-right-from-square shrink-0 text-2xs leading-4"
                    role="img"
                  />
                </a>
              </div>
            </div>
            <form
              class="w-full"
            >
              <div
                class=""
              >
                <div
                  class="input input--select    "
                  data-testid="select"
                >
                  <label
                    class="!text-xs !translate-y-0"
                    for="File source"
                  >
                    File source
                  </label>
                  <div
                    class=" css-b62m3t-container"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-5-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="input-select__control css-13cymwt-control"
                    >
                      <div
                        class="input-select__value-container input-select__value-container--has-value css-1fdsijx-ValueContainer"
                      >
                        <span
                          class="text-sm text-neutral-400 mr-1"
                        >
                          Raw YAML
                        </span>
                        <input
                          aria-autocomplete="list"
                          aria-expanded="false"
                          aria-haspopup="true"
                          aria-readonly="true"
                          class="css-mohuvp-dummyInput-DummyInput"
                          id="File source"
                          inputmode="none"
                          role="combobox"
                          tabindex="0"
                          value=""
                        />
                      </div>
                      <div
                        class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                    <input
                      name="File source"
                      type="hidden"
                      value="YAML"
                    />
                  </div>
                  <input
                    name="File source"
                    type="hidden"
                    value="YAML"
                  />
                  <div
                    class="absolute top-1/2 -translate-y-1/2 right-4 pointer-events-none"
                  >
                    <span
                      class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
                      role="img"
                    />
                  </div>
                </div>
              </div>
              <section
                class="flex flex-col"
              >
                <h2
                  class="text-neutral-400 text-base font-bold mt-10 mb-2"
                >
                  Override with raw Yaml
                </h2>
                <p
                  class="text-sm text-neutral-350 mb-6"
                >
                  You can define here the YAML containing the overrides you want to apply. The YAML will be stored by Qovery and can be updated later within the settings but no history will be retained.
                </p>
                <div
                  class="flex flex-col gap-3"
                >
                  <div
                    class="border border-neutral-250 bg-neutral-100 rounded mb-5 w-full"
                    data-testid="block-content"
                  >
                    <div
                      class="flex items-center justify-between h-9 px-4 border-b border-neutral-250"
                    >
                      <h2
                        class="font-medium text-neutral-400 text-ssm"
                      >
                        Raw YAML
                      </h2>
                      <button
                        class="inline-flex items-center shrink-0 font-medium transition text-xs px-1 py-0.5 rounded hover:text-neutral-400"
                        type="button"
                      >
                        <span
                          class="icon-solid-pen shrink-0 "
                          role="img"
                        />
                      </button>
                    </div>
                    <div
                      class="p-0"
                    >
                      <section
                        style="display: flex; position: relative; text-align: initial; width: 100%; height: 300px;"
                      >
                        <div
                          style="display: flex; height: 100%; width: 100%; justify-content: center; align-items: center;"
                        >
                          <div
                            class="aspect-square border-2  rounded-full border-solid border-r-neutral-500 animate-spin w-4 "
                            data-testid="spinner"
                          />
                        </div>
                        <div
                          style="width: 100%; display: none;"
                        />
                      </section>
                    </div>
                  </div>
                </div>
              </section>
              <div
                class="flex justify-between mt-10"
              >
                <button
                  class="inline-flex items-center shrink-0 font-medium transition text-sm h-10 px-4 rounded border border-neutral-250 dark:border-neutral-350 text-neutral-400 dark:text-neutral-300 hover:[&:not(:active)]:border-neutral-300 active:bg-neutral-150 disabled:text-neutral-300 disabled:bg-neutral-150 disabled:border-none bg-neutral-100 dark:bg-neutral-500 dark:hover:bg-neutral-550"
                  type="button"
                >
                  Back
                </button>
                <div
                  class="flex gap-3"
                >
                  <button
                    class="inline-flex items-center shrink-0 font-medium transition text-sm h-10 px-4 rounded bg-brand-500 hover:bg-brand-600 text-white disabled:text-brand-300 disabled:bg-brand-100 disabled:border-none"
                    type="submit"
                  >
                    Continue
                  </button>
                </div>
              </div>
            </form>
          </section>
        </div>
      </section>
      <aside
        class="w-[30%] sticky top-0 pt-14 pl-10 hidden lg:block"
      >
        <div
          class="bg-neutral-50 p-8 border-neutral-200 border-b border-l rounded max-w-[22.5rem] "
          data-testid="funnel-body-help"
        >
          <section
            class="flex flex-col"
          >
            <svg
              class="shrink-0 w-7 h-7 mb-4"
              fill="none"
              height="28"
              name="INFORMATION"
              viewBox="0 0 28 28"
              width="1.5rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="#B160F0"
                height="28"
                rx="3"
                width="28"
              />
              <path
                d="M11.813 19.191c0 .165.027.329.136.465l.465.711c.137.219.465.383.738.383h1.668c.274 0 .602-.164.739-.383l.464-.71c.082-.11.137-.329.137-.466l.027-1.066h-4.375v1.066zm2.16-12.414c-2.79 0-4.786 2.243-4.786 4.785 0 1.204.438 2.297 1.176 3.145.465.52 1.149 1.613 1.422 2.516v.027h4.402v-.027c.274-.903.957-1.996 1.422-2.516a4.673 4.673 0 001.203-3.145c0-2.652-2.187-4.812-4.84-4.785zM14 9.402c-1.23 0-2.188.957-2.188 2.16a.45.45 0 01-.437.438.432.432 0 01-.438-.438c0-1.667 1.368-3.035 3.063-3.035a.47.47 0 01.438.438.45.45 0 01-.438.437z"
                fill="#fff"
              />
            </svg>
            <h2
              class="text-neutral-400 text-base font-bold mb-4"
            >
              Using values overrides
            </h2>
            <ul
              class="list-disc pl-4 text-neutral-400 text-sm mb-8"
            >
              <li
                class="mb-3"
              >
                Your helm chart might have already a variables.yaml file with some basic configuration. In this section you can define your own overrides to customize the helm chart behaviour.
              </li>
              <li
                class="mb-3"
              >
                You can define the overrides by selecting a YAML file from a git repository (preferred) or by passing a raw YAML file.
              </li>
              <li
                class="mb-3"
              >
                You can use the Qovery environment variables as overrides by using the placeholder "qovery.env.ENV_VAR_NAME” (Example: qovery.env.DB_URL. Qovery will manage the replacement of those placeholders at deployment time.
              </li>
              <li
                class="mb-3"
              >
                To get all the Qovery functionalities, add the macro “qovery.labels.service” and "qovery.annotations.service" within the field managing the labels/annotations assigned to the deployed Pods/Deployments/Services/Jobs. Qovery will automatically replace the macro with a dedicated set of labels/annotations.
              </li>
            </ul>
            <div
              class="py-8 border-t border-neutral-200 -mx-8 px-8 pb-0"
              data-testid="help-section"
            >
              <p
                class="text-neutral-400 text-sm mb-5"
              >
                Need help? You may find these links useful
              </p>
              <div
                class="flex flex-col gap-2"
              >
                <a
                  class="cursor-pointer transition duration-100 font-medium items-center gap-1 text-sky-500 hover:text-sky-600 text-sm flex"
                  href="https://hub.qovery.com/docs/using-qovery/configuration/helm/"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  How to configure my Helm chart
                  <span
                    class="icon-solid-arrow-up-right-from-square shrink-0 text-xs leading-5"
                    role="img"
                  />
                </a>
              </div>
            </div>
          </section>
        </div>
      </aside>
    </div>
  </div>
  <script
    src="https://cdn.jsdelivr.net/npm/monaco-editor@0.43.0/min/vs/loader.js"
  />
</body>
`;
